Add `Sys.which()`, use that to find `curl` in `download()`

staticfloat opened the pull request:
It's better to invoke `curl` directly (e.g. via `curl --help`) in order to see if it exists than to use `which curl` and see if that successfully passes.  This fixes test errors on e.g. our CentOS 7.3 buildbots, which do not have `which` available by default.

staticfloat committed:
Don't use `which curl`, use `Sys.which("curl")`

This fixes issues on systems that don't have `which` available

iamed2 commented:
Maybe someone with FreeBSD can confirm, but I don't think `fetch` supports `--help` or any other no-op argument.

ararslan commented:
Can confirm, it does not.

stevengj commented:
Might be nice to have a portable `which`, analogous to Python's `distutils.spawn.find_executable` or `shutils.which`, if we don't have one already.

ararslan commented:
@stevengj, Elliot and I were just talking about that offline. He intends to add such a function based off of https://github.com/ararslan/BuildUtils.jl/blob/master/src/BuildUtils.jl#L46-L74. :slightly_smiling_face:

staticfloat commented:
Yep, I'll tack it onto this PR, just writing up the tests now.  It'll be called `Sys.which()`, and I expect a full course of bikeshedding.  Don't let me down guys.

staticfloat committed:
Add `Sys.which(program_name::AbstractString)`

This acts as a julia-native version of the `which` command found on most
*nix systems; it searches the path for an executable of the given name,
returning the absolute path if it exists, and throwing an
`ArgumentError()` if it does not.

ararslan commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,66 @@ windows_version
+    # If prog has a slash, we know the user wants to determine whether the given
+    # file exists and is executable, and to not search the `PATH`.  Note that
+    # Windows can have either `\\` or `/` in its paths:
+    dirseps = @static if is_windows() ["/", "\\"] else ["/"] end
```
`is_windows` is deprecated to `iswindows`

ararslan commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,66 @@ windows_version
+    # Windows can have either `\\` or `/` in its paths:
+    dirseps = @static if is_windows() ["/", "\\"] else ["/"] end
+
+    if any(contains.(program_name, dirseps))
```
`contains(a, b)` is deprecated to `occursin(b, a)`

ararslan commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,66 @@ windows_version
+    # We use `access()` and `X_OK` to determine if a given path is executable
+    # by the current user.  `X_OK` comes from `unistd.h`.
+    X_OK = 1 << 0
+    can_execute(program_name) = @static if is_windows()
```
This could be moved out of the function

stevengj commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,81 @@ windows_version
+        ["/", "\\"]
+    else
+        ["/"]
+    end
```
Probably a tuple of chars here rather than an array of strings?

staticfloat replied:
Are tuples preferred here for performance considerations?

stevengj commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,81 @@ windows_version
+    # If we have been given just a program name (not a relative or absolute
+    # path) then we should search `PATH` for it here:
+    path = get(ENV, "PATH", "")
+    pathsep = @static if iswindows()
```
`@static iswindows() ? ';' : ':'` would be shorter and probably more readable for conditionals like this.

staticfloat replied:
Ah, for some reason I thought `@static` didn't work with the ternary operator.  Nice.

vtjnash replied:
The `@static` isn't entirely necessary

stevengj commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,81 @@ windows_version
+
+    # On windows, we need to check both $(program_name) and $(program_name).exe
+    program_matches(file, program_name) = @static if iswindows()
+        file == program_name || file == "$(program_name).exe"
```
Don't you need to check everything in `get(ENV, "PATHEXT", "exe")`

stevengj replied:
Also, you presumably only want to check appending an extension if `program_name` doesn't already have an extension.

vtjnash replied:
No to the first (also `.cmd`), yes to the second (with qualifications). We should probably try to return the same answer as our underlying spawn API (https://github.com/libuv/libuv/blob/edf05b97f03472def2837ceb2d6bf61a0d0dc248/src/win/process.c#L299)

stevengj commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,81 @@ windows_version
+        file == program_name
+    end
+
+    for dir in split(path, pathsep), file in readdir(dir)
```
Why walk the directory tree?  This seems very inefficient if the directories contain lots of entries, as will be common.  Instead, you could do something like:
```jl
for dir in split(path, pathsep)
     file = joinpath(path, program_name)
     isfile(file) && _can_execute(file) && return abspath(file)
     .... check extensions on windows ....
end
```
right?

stevengj commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,81 @@ windows_version
+
+Returns `true` if the given `path` has executable permissions.
+"""
+function _can_execute(path::AbstractString)
```
`isexecutable` seems more idiomatic.

stevengj commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,81 @@ windows_version
+        ["/"]
+    end
+
+    if any(occursin.(dirseps, program_name))
```
or just
```jl
if occursin(@static(iswindows() ? r"[/\\]" : '/'), program_name)
```

vtjnash commented on base/sysinfo.jl:
```diff
@@ -323,6 +323,22 @@ windows_version
+    # by the current user.  `X_OK` comes from `unistd.h`.
+    X_OK = 1 << 0
+    @static if iswindows()
+        ccall(:_access, Cint, (Ptr{UInt8}, Cint), path, X_OK) == 0
```
`access` on windows doesn't have an `X_OK` parameter

staticfloat replied:
Ah, that's an excellent point.  I just spent half an hour trying to parse MSDN and figure out how `GetNamedSecurityInfo` can help me figure out the right ACL values to see if something is executable by the current user, but it defeated me.  I'm just going to return `true` on Windows.

vtjnash replied:
Yes, the `which` function on Windows probably shares fairly little with the same function on posix. I expect they likely should just be completely independent implementations

staticfloat replied:
.........`busybox-w32` uses `access(path, X_OK)`.  Sigh.

vtjnash replied:
I’m not too surprised. It’s a posix emulation later, probably intended to approximate the result you’d get on `busybox sh` or cygwin (similar to how the real which program on posix approximates the likely answer)

vtjnash commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,83 @@ windows_version
+    pathsep = @static iswindows() ? ';' : ':'
+
+    # On windows, if the program name does not already have an extension,
+    # we need to tack all the values in PATHEXT onto the end and check for
```
Nope, see how libuv implements this function. IIRC, the distinction is that PATHEXT is for DOS, not the Shell32.

staticfloat replied:
Which function are you referring to? What should I look for in `libuv`?

I'm not sure what you mean in stating that `PATHEXT` is only for DOS; when I try to `run()` a command, I expect it to work like I've typed it into `cmd.exe`, and `cmd.exe` does pay attention to `PATHEXT`.

vtjnash replied:
See my libuv link in the other comment thread

stevengj replied:
You want to match the semantics of `search_path` in libuv's `src/win/process.c`, I guess, since this is what `run` uses.

Unfortunately, the `search_path` function is not exported.  But since we use our own libuv fork anyway, maybe we should just modify it to export a `uv_win_search_path` function on Windows?

stevengj replied:
Libuv says:
> Furthermore, it tries to follow the semantics that cmd.exe, with this exception that PATHEXT environment variable isn't used. Since CreateProcess can start only .com and .exe files, only those extensions are tried. This  behavior equals that of msvcrt's spawn functions.

stevengj replied:
Note also that libuv writes *CMD does not trim leading/trailing whitespace from path/pathex entries*

stevengj commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,83 @@ windows_version
+    program_names = [program_name]
+    @static if iswindows()
+        if !isempty(splitext(program_name)[2])
+            pathexts = get(ENV, "PATHEXT", ".com; .exe; .bat; .cmd")
```
no spaces in the default

stevengj commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,83 @@ windows_version
+    @static if iswindows()
+        if !isempty(splitext(program_name)[2])
+            pathexts = get(ENV, "PATHEXT", ".com; .exe; .bat; .cmd")
+            pathexts = strip.(split(pathexts, ';'))
```
Is the `strip` necessary?  Can PATHEXT contain whitespace?

staticfloat commented:
So looking through this, I guess we have to make the decision; do we want to behave like `cmd` here, or do something else?  Since the most logical use of this function is to be used with `run()`, which I assume is going to eventually sub off to `CreateProcess`, we should probably ignore `PATHEXT` completely just like libuv does, and only search for `.exe` and `.com` files.

stevengj commented:
I think `which("foo")` should return the same program that ```run(`foo`)``` would execute, which means it needs to follow libuv.

stevengj commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,91 @@ windows_version
+"""
+function isexecutable(path::AbstractString)
+    @static if iswindows()
+        return true
```
`isfile(path)`?

stevengj commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,91 @@ windows_version
+        for program_name in program_names
+            program_path = joinpath(path_dir, program_name)
+            # If we find something that matches our name and we can execute
+            if isfile(program_path) && isexecutable(program_path)
```
If you change `isexecutable` to `isfile` on Windows as suggested above, then you can omit the `isfile` check here.

staticfloat commented:
Okay, I think I've addressed all the comments so far.

ararslan commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,91 @@ windows_version
+"""
+function isexecutable(path::AbstractString)
+    @static if iswindows()
+        return isfile(path)
```
Can't you determine whether it has executable permissions using `_access` on Windows?

staticfloat replied:
Nope.  Windows doesn't have an `X_OK` bit.  :(

https://msdn.microsoft.com/en-us/library/1w06ktdy.aspx

ararslan replied:
Ah okay, sorry for the noise then.

stevengj commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,91 @@ windows_version
+        # If this file does not even exist, fail out
+        if !isfile(program_name)
+            throw(ArgumentError("$program_name does not exist"))
+        end
```
this check is now redundant with `isexecutable`

stevengj commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,91 @@ windows_version
+    # Windows can have either `\\` or `/` in its paths:
+    dirseps = @static iswindows() ? ('/', '\\') : ('/', )
+
+    if any(occursin.(dirseps, program_name))
```
`if occursin('/', program_name) || @static(iswindows() ? occursin('\\', program_name) : false)` would avoid the broadcast allocation, not that it matters much here.

staticfloat replied:
Agreed that it doesn't matter much here, and I like the readability of what I have here already.

vtjnash replied:
Or `:` – if present, it's a relative path with search disabled

staticfloat replied:
I don't understand what you mean, Jameson.  Could you explain?

stevengj commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,91 @@ windows_version
+
+    # On windows we always check the current directory as well
+    @static if iswindows()
+        insert!(path_dirs, 1, pwd())
```
`pushfirst!` would be clearer

stevengj commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,91 @@ windows_version
+    program_names = AbstractString[]
+    @static if iswindows()
+        # If the file already has an extension, try that name first
+        if !isempty(splitext(program_name)[2])
```
`occursin(Filesystem.path_ext_splitter, program_name)`

staticfloat replied:
That doesn't seem to work:

```
julia> occursin(Base.Filesystem.path_ext_splitter, "")
false

julia> occursin(Base.Filesystem.path_ext_splitter, "e")
true
```

stevengj commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,91 @@ windows_version
+absolute path is passed in for `program_name`, that exact path is tested for
+executable permissions only, no searching of `PATH` is performed.
+"""
+function which(program_name::AbstractString)
```
Probably 
```jl
which(program_name::AbstractString) = which(String(program_name))
function which(program_name::String)
    ...
end
```
as otherwise `program_name` will get converted to `String` multiple times if it is some other string type.

stevengj commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,91 @@ windows_version
+    end
+
+    # Build a list of program names that we're going to try
+    program_names = AbstractString[]
```
`String[]`

vtjnash commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,88 @@ windows_version
+        end
+
+        # If it all checks out, return the abspath
+        return abspath(program_name)
```
The Window's rules aren't suppose to short-circuit like this – the `program_names` rules still apply in this case

vtjnash commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,88 @@ windows_version
+    end
+
+    # If we couldn't find anything, complain
+    throw(ArgumentError("$program_name not found"))
```
not an ArgumentError

staticfloat replied:
What would you call it?

stevengj replied:
Probaby just `error`?

`open` on a non-existent file throws `SystemError`, but that's not appropriate here since `errno` was not set.

vtjnash commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,82 @@ windows_version
+            program_path = joinpath(path_dir, program_name)
+            # If we find something that matches our name and we can execute
+            if isexecutable(program_path)
+                return abspath(program_path)
```
`abspath` doesn't compute the same location as `isexecutable`. Did you mean `realpath` (which would match the binutils `which` program) or to call `abspath` first (which would be odd)?

staticfloat replied:
You're right; the `abspath()` was vestigial from an older way of doing things; now that `program_path` is a `joinpath()` output, we never have the case where we need to turn a relative path into an absolute one; this path is always absolute.

vtjnash commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,82 @@ windows_version
+
+    for path_dir in path_dirs
+        for program_name in program_names
+            program_path = joinpath(path_dir, program_name)
```
Isn't this call attempting to do `joinpath(dirname(program_name), program_name)`?

staticfloat replied:
That is what it does if `program_name` is `foo/bar.exe`, but if the name is `bar.exe`, then `path_dir` will be things like `$(pwd)` and the elements of `%PATH%`.  (To use windows as an example)

ggggggggg commented:
Why throw an error if the path isn't found, why not just return `nothing`? `which foo` at the command line just doesn't print, it doesn't complain. Likewise `Base` functions like `find*` don't error if they fail to find a match.

samoconnor commented:
A bit off topic:
HTTP.jl implements redirect and HTTP/HTTPs proxying (not tested much yet).
HTTP.jl does not have a `download` function, but it has `open ... do ...` 
```julia
HTTP.open("GET", "http://foobar.com/abc.zip") do io
    write("abc.zip", io)
end
```

Are there important curl features that HTTP.jl lacks?
Is there a plan to eventually put HTTP.jl into StdLib?
Would adding HTTP.jl to StdLib be more palatable if the sever functionality was split into a seperate package?

stevengj commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,83 @@ windows_version
+for executable permissions only (with `.exe` and `.com` extensions added on
+Windows platforms), no searching of `PATH` is performed.
+"""
+function which(program_name::String)
```
@musm, see my comment above: https://github.com/JuliaLang/julia/pull/26559#pullrequestreview-108491386

staticfloat commented on base/sysinfo.jl:
```diff
@@ -321,4 +323,83 @@ windows_version
+function which(program_name::String)
+    # Build a list of program names that we're going to try
+    program_names = String[]
+    @static if iswindows()
```
Why is the static unnecessary? There’s no reason to do this work on non-windows systems.

stevengj replied:
The main reason for `@static`, in my understanding, is for something like a `ccall` that might not compile at all on some platforms, or for complex predicates that can't be inlined and optimized away.

staticfloat replied:
Excellent, I usually don't trust compilers to hold a spoon from the right end, but this looks great.

staticfloat committed:
Add `NEWS.md` entry for `Sys.which()`

staticfloat commented:
Error is documenter-related

ararslan commented on base/sysinfo.jl:
```diff
@@ -322,4 +324,83 @@ windows_version
+    end
+
+    # If we couldn't find anything, complain
+    error("$program_name not found")
```
This should probably be an `ArgumentError`.

ararslan replied:
Ah, I see it originally was but Jameson said it shouldn't be. Just using `error` here seems unfortunately general.

staticfloat replied:
Yeah, but I don't have a good feel for what errors should be what, so in the absence of a better idea, I'm going to keep it as `error()` for now.

ararslan commented on test/spawn.jl:
```diff
@@ -536,3 +536,77 @@ let
+end
+
+mktempdir() do dir
+    pathsep = @static if Sys.iswindows() ";" else ":" end
```
A ternary would be less weird here

staticfloat replied:
`pathsep = Sys.iswindows() ? ";" : ":"` is too much punctuation for me.

ararslan commented on test/spawn.jl:
```diff
@@ -536,3 +536,77 @@ let
+            bar_path = joinpath(dir, "bin1", "bar.exe")
+            touch(bar_path)
+            chmod(bar_path, 0o777)
+            cd("$(dir)") do
```
`dir` is already a `String`; it doesn't need to be interpolated.

ararslan commented on test/spawn.jl:
```diff
@@ -536,3 +536,77 @@ let
+
+        @static if Sys.iswindows()
+            # On windows, check that pwd() takes precedence, except when we provide a path
+            cd("$(dir)/bin2") do
```
`joinpath(dir, "bin2")`

ararslan commented on base/sysinfo.jl:
```diff
@@ -322,4 +324,83 @@ windows_version
+"""
+    Sys.isexecutable(path::String)
+
+Returns `true` if the given `path` has executable permissions.
```
We word our docs in the imperative, i.e. this should be "return" rather than "returns." Same for the `which` docstring below.

staticfloat replied:
Hah, I was basing this off of the `Sys.windows_version()` docstring above it.  I'll fix that too.

ararslan commented on base/sysinfo.jl:
```diff
@@ -23,7 +23,8 @@ export BINDIR,
        isunix,
-       iswindows
+       iswindows,
+       which
```
Maybe export `isexecutable` as well?

staticfloat committed:
Fix wording of `Sys.windows_version()` to be in the imperative

https://github.com/JuliaLang/julia/pull/26559#discussion_r187183056

staticfloat commented:
This passes all CI now.  Win32 build failure is unrelated, the `spawn` test passes, which is the only thing that matters for this PR.

Absent any other comments, I'm going to merge this tomorrow.

staticfloat commented:
I intentionally didn't; each commit is intended to stand alone.
